scripts/collect-npm-stats.ts

```diff
@@ -0,0 +1,166 @@
+  const url = `https://api.npmjs.org/versions/${encodedPackage}/last-week`;
+
+  console.log(`Fetching stats for ${packageName}...`);
+
+  const response = await fetch(url);
+  if (!response.ok) {
+    throw new Error(
+      `Failed to fetch stats for ${packageName}: ${response.status} ${response.statusText}`,
+    );
+  }
+
+  return response.json() as Promise<NpmDownloadResponse>;
+}
+
+function aggregateByMajorVersion(downloads: Record<string, number>): Record<string, number> {
+  const majorVersions: Record<string, number> = {};
+
+  for (const [version, count] of Object.entries(downloads)) {
+    const major = semver.major(version);
+    const majorKey = major.toString();
+
+    if (!majorVersions[majorKey]) {
+      majorVersions[majorKey] = 0;
+    }
+    majorVersions[majorKey] += count;
+  }
+
+  return majorVersions;
+}
+
+async function readExistingData(filePath: string): Promise<HistoricalData | null> {
+  try {
+    const content = await readFile(filePath, 'utf-8');
+    return JSON.parse(content);
+  } catch (error) {
+    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
+      return null;
+    }
+    throw error;
+  }
+}
+
+async function updateHistoricalData(
+  packageName: string,
+  newDownloads: Record<string, number>,
+  existingData: HistoricalData | null,
+): Promise<HistoricalData> {
+  const timestamp = Date.now();
+
+  if (!existingData) {
+    // Create new data structure
+    const downloads: Record<string, number[]> = {};
+    for (const [major, count] of Object.entries(newDownloads)) {
+      downloads[major] = [count];
+    }
+
+    return {
+      package: packageName,
+      timestamps: [timestamp],
+      downloads,
+    };
+  }
+
+  // Update existing data
+  const updatedData: HistoricalData = {
+    ...existingData,
+    timestamps: [...existingData.timestamps, timestamp],
+    downloads: { ...existingData.downloads },
+  };
+
+  // Add new download counts
+  for (const [major, count] of Object.entries(newDownloads)) {
+    if (!updatedData.downloads[major]) {
+      // New major version - backfill with zeros for historical timestamps
+      updatedData.downloads[major] = new Array(existingData.timestamps.length).fill(0);
+    }
+    updatedData.downloads[major].push(count);
+  }
+
+  // Ensure all existing major versions have a new entry (fill with 0 if no downloads)
+  for (const major of Object.keys(existingData.downloads)) {
+    if (!newDownloads[major]) {
+      updatedData.downloads[major].push(0);
+    }
+  }
+
+  return updatedData;
+}
+
+async function processPackage(packageName: string): Promise<void> {
+  try {
+    // Fetch current stats
+    const stats = await fetchPackageStats(packageName);
+
+    // Aggregate by major version
+    const aggregatedDownloads = aggregateByMajorVersion(stats.downloads);
+
+    // Determine file path
+    const dataDir = join(process.cwd(), 'data', 'npm-versions');
+    const filePath = join(dataDir, `${packageName}.json`);
```

Comment:
[Jsonl](https://www.google.com/search?q=jsonl+file) file format instead of json seems more apt for this use case.
No need of reading the existing contents and merging the new data and writing back.
You just append the new json string as a line at the end of the file.

Reply:
Jsonl is optimal for reading as well since we can read it line by line for the data points.
But I agree that given its weekly, it doesn't make much sense for early optimization.